File: ig-codegen/src/main/java/io/github/dizuker/igcodegen/FhirResourceSummary.java
package io.github.dizuker.igcodegen;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import java.util.List;
import org.jspecify.annotations.Nullable;

/** The subset of a FHIR resource JSON file's fields needed to classify it. */
@JsonIgnoreProperties(ignoreUnknown = true)
public record FhirResourceSummary(
    @Nullable String resourceType,
    @Nullable String id,
    @Nullable String url,
    @Nullable String version,
    @Nullable String kind,
    @Nullable String derivation,
    @Nullable String type,
    @Nullable String content,
    @Nullable List<Concept> concept,
    @Nullable Snapshot snapshot,
    @Nullable Compose compose) {

  /** A CodeSystem.concept entry; {@code concept} holds nested child concepts, if any. */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record Concept(
      @Nullable String code, @Nullable String display, @Nullable List<Concept> concept) {}

  /** A StructureDefinition.snapshot, reduced to what's needed to inspect {@code value[x]}. */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record Snapshot(@Nullable List<Element> element) {}

  /**
   * A snapshot element, reduced to what's needed to inspect {@code Extension.value[x]} and, for a
   * {@code CodeableConcept}/{@code Coding}-typed value, whether its {@code system} is pinned to one
   * fixed CodeSystem, either directly ({@code fixedUri}) or via a {@code required} {@code binding}
   * to a ValueSet that itself draws from a single CodeSystem.
   */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record Element(
      @Nullable String path,
      @Nullable String max,
      @Nullable List<ElementType> type,
      @Nullable String fixedUri,
      @Nullable Binding binding) {}

  /** One entry of an element's {@code type} array, e.g. {@code {"code": "CodeableConcept"}}. */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record ElementType(@Nullable String code) {}

  /** An element's terminology binding, e.g. {@code {"strength": "required", "valueSet": "..."}}. */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record Binding(@Nullable String strength, @Nullable String valueSet) {}

  /** A ValueSet.compose, reduced to what's needed to tell whether it draws from one CodeSystem. */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record Compose(
      @Nullable List<ComposeInclude> include, @Nullable List<ComposeInclude> exclude) {}

  /**
   * One entry of {@code ValueSet.compose.include} (or {@code .exclude}). {@code valueSet} holds
   * canonical references to other ValueSets being imported wholesale, as opposed to {@code system}
   * pulling concepts directly from one CodeSystem.
   */
  @JsonIgnoreProperties(ignoreUnknown = true)
  public record ComposeInclude(
      @Nullable String system,
      @Nullable List<Object> concept,
      @Nullable List<Object> filter,
      @Nullable List<String> valueSet) {}
}

File: ig-codegen-python/src/ig_codegen/fhir_resource_summary.py
"""The subset of a FHIR resource JSON file's fields needed to classify it."""

from __future__ import annotations

from dataclasses import dataclass, field


@dataclass(frozen=True)
class Concept:
    """A CodeSystem.concept entry; ``concept`` holds nested child concepts, if any."""

    code: str | None = None
    display: str | None = None
    concept: list[Concept] = field(default_factory=list)

    @staticmethod
    def from_json(data: dict) -> Concept:
        return Concept(
            code=data.get("code"),
            display=data.get("display"),
            concept=[Concept.from_json(c) for c in data.get("concept") or []],
        )


@dataclass(frozen=True)
class FhirResourceSummary:
    resource_type: str | None = None
    id: str | None = None
    url: str | None = None
    version: str | None = None
    kind: str | None = None
    derivation: str | None = None
    type: str | None = None
    content: str | None = None
    concept: list[Concept] = field(default_factory=list)

    @staticmethod
    def from_json(data: dict) -> FhirResourceSummary:
        return FhirResourceSummary(
            resource_type=data.get("resourceType"),
            id=data.get("id"),
            url=data.get("url"),
            version=data.get("version"),
            kind=data.get("kind"),
            derivation=data.get("derivation"),
            type=data.get("type"),
            content=data.get("content"),
            concept=[Concept.from_json(c) for c in data.get("concept") or []],
        )
